api/v1beta3/vcdcluster_webhook.go

```diff
@@ -61,8 +65,26 @@ func (r *VCDCluster) ValidateCreate() (admission.Warnings, error) {
 func (r *VCDCluster) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
 	vcdclusterlog.Info("validate update", "name", r.Name)
 
-	// TODO(user): fill in your validation logic upon object update.
-	return nil, nil
+	var allErrs field.ErrorList
+
+	oldVCDCluster, ok := old.(*VCDCluster)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a VCDCluster but got a %T", old))
+
+	}
+	if !cmp.Equal(oldVCDCluster.Spec.ControlPlaneEndpoint, APIEndpoint{}) &&
```

Comment:
This looks like we will allow an empty APIEndpoint to be updated. Does the patch also go through this hook? I am worried about the case where:
1. Someone applies a capi yaml
2. VCDCluster Controller picks it up and is still trying to figure out an free IP
3. Someone applies a control-plane endpoint to the object
    a. This succeeds based on the above logic
4. When the controller tries to update it, the update fails.

This is fine. However, we need to do a couple of more steps after 4
1. VCDCluster controller must update the IP into the spec before actually using it.
2. If the update fails, VCDCluster returns with an error and gets into the retry loop.
3. In the retry loop, the controller should detect the user-specified IP and use it.

Can you confirm that this would be the workflow?

Reply:
Looks good to me from IP reservation point of view. The claim marker will make sure that the IP is not leaked and deleted when the LB is destroyed.